### PR TITLE
pkg/gitignore: read gitignore file and extract patterns

### PR DESCRIPTION
**Pull request by aleh-null**
Port of `"github.com/go-git/go-git/plumbing/format/gitignore"` for `afero.Fs`. 

ATM I don't think we need to handle: 
- Global patterns in user's ~/.gitconfig  
- System's patterns in /etc/gitconfig

### CODE REVIEW

**Review by artmoskvin — COMMENTED** [review]
Looks awesome! One small question.

File: pkg/gitignore/dir.go [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+	gitignoreFile   = ".gitignore"
+	gitconfigFile   = ".gitconfig"
+	systemFile      = "/etc/gitconfig"
+	infoExcludeFile = gitDir + "/info/exclude"
+)
+
+// ReadPatterns reads gitignore patterns recursively traversing through the directory
+// structure. The result is in the ascending order of priority (last higher).
+func ReadPatterns(fs afero.Fs, path []string) ([]gitignore.Pattern, error) {
+	path = path[:len(path):len(path)]
```
[the remark this answers is not in the document]
Comment (aleh-null) [reply pkg/gitignore/dir.go:28]: Basically to ensure that slice of this slice on append does not write to underlying shared array but creates a new one. https://go.dev/play/p/jZdNBg_BMbB

**Review by artmoskvin — APPROVED** [review]

File: pkg/gitignore/dir.go [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+	}
+
+	return ps, nil
+}
+
+// readIgnoreFile reads a specific .gitignore file
+func readIgnoreFile(fs afero.Fs, path []string, ignoreFile string) (ps []gitignore.Pattern, err error) {
+	path = path[:len(path):len(path)]
+
+	f, err := fs.Open(filepath.Join(append(path, ignoreFile)...)) // this might cause an issue
```
Comment (artmoskvin) [inline pkg/gitignore/dir.go:73]: is the comment still relevant?

File: pkg/gitignore/dir.go [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+	if err != nil {
+		// .gitignore does not exist along this path
+		if os.IsNotExist(err) {
+			return nil, nil
```
Comment (artmoskvin) [inline pkg/gitignore/dir.go:77]: why do we return `nil` instead of `err`?

File: pkg/gitignore/dir.go (hunk 2 above)
Comment (aleh-null) [reply pkg/gitignore/dir.go:73]: nope.

File: pkg/gitignore/dir.go (hunk 3 above)
Comment (aleh-null) [reply pkg/gitignore/dir.go:77]: we try to read .ignorefile, if it does not exist we will not have any patterns.

File: pkg/gitignore/dir.go [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+	return ps, nil
+}
+
+func readDir(fs afero.Fs, dir string) ([]os.FileInfo, error) {
+	d, err := fs.Open(dir)
+	if err != nil {
+		return nil, err
+	}
+
+	return d.Readdir(0)
```
Comment (artmoskvin) [inline pkg/gitignore/dir.go:103]: what's 0? what does it mean?

Comment (aleh-null) [reply pkg/gitignore/dir.go:103]: it's a count that you can specify in the afero implementation. Reads all if set to zero.

Comment (aleh-null) [reply pkg/gitignore/dir.go:103]: https://github.com/artmoskvin/hide/blob/8a2c595dd54dbe586de18a04cdb5e1ed63a5f992/vendor/github.com/spf13/afero/mem/file.go#L160

File: pkg/gitignore/dir.go (hunk 3 above)
Comment (artmoskvin) [reply pkg/gitignore/dir.go:77]: ah ok, I see now. if the file doesn't exist, we move on and it's not an error.

**Review by artmoskvin — APPROVED** [review]

**Pull request merged**